main_test.go

```diff
@@ -0,0 +1,439 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/thriftrw/thriftrw-go/compile"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDeterminePackagePrefix(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "thriftrw-main-test")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	genDir, err := filepath.Abs("gen")
+	require.NoError(t, err)
+
+	tests := []struct {
+		around func(func())
+		dir    string
+		result string
+		errMsg string
+	}{
+		{
+			around: func(f func()) {
+				gopath := os.Getenv("GOPATH")
+				require.NoError(t, os.Unsetenv("GOPATH"))
+				defer os.Setenv("GOPATH", gopath)
+
+				f()
+			},
+			dir:    genDir,
+			errMsg: "$GOPATH is not set",
+		},
+		{
+			dir:    tmpDir,
+			errMsg: "not inside $GOPATH",
+		},
+		{
+			dir:    genDir,
+			result: "github.com/thriftrw/thriftrw-go/gen",
+		},
+		{
+			around: func(f func()) {
+				gopath := os.Getenv("GOPATH")
+				require.NoError(t, os.Setenv("GOPATH", tmpDir+":"+gopath))
+				defer os.Setenv("GOPATH", gopath)
+			},
```

Comment:
no call to `f`

the `around` functionality seems too generic and is more error-prone to bugs like this. suggest just adding a property like,
```go
  overrideGoPath func(curGoPath string) string
```